File: csharp/api/utils/TimeConverter.cs
namespace api.utils;

using Google.Protobuf.WellKnownTypes;
using System;

public class TimeConverter
{
    // Convert from Google.Protobuf.Timestamp to System.DateTime
    public static DateTime? ToDateTime(Timestamp? protoTimestamp)
    {
        if (protoTimestamp == null)
        {
            return null;
        }

        return DateTimeOffset.FromUnixTimeSeconds(protoTimestamp.Seconds)
            .AddTicks(protoTimestamp.Nanos / 100)
            .UtcDateTime;
    }

    // Convert from System.DateTime to Google.Protobuf.Timestamp
    public static Timestamp? ToProtobufTimestamp(DateTime? dateTime)
    {
        if (dateTime == null)
        {
            return null;
        }

        DateTimeOffset dto = new DateTimeOffset(dateTime.Value.ToUniversalTime());
        return new Timestamp
        {
            Seconds = dto.ToUnixTimeSeconds(),
            Nanos = dto.Millisecond * 1_000_000
        };
    }
}

File: java-db/src/main/java/via/dk/util/TimeConverter.java
package via.dk.util;

import java.sql.Timestamp;

/**
 * Static class for converting timestamps between Java and Protobuf.
 */
public class TimeConverter
{
	/**
	 * Converts a Protobuf timestamp to a Java timestamp.
	 * @param protoTimestamp The Protobuf timestamp to convert.
	 * @return The converted Java timestamp.
	 */
  public static Timestamp toJavaTimestamp(com.google.protobuf.Timestamp protoTimestamp) {
    if (protoTimestamp == null) {
      return null;
    }

    long seconds = protoTimestamp.getSeconds();
    int nanos = protoTimestamp.getNanos();
    long millis = seconds * 1000 + nanos / 1_000_000;
    return new Timestamp(seconds * 1000 + nanos / 1_000_000);
  }

	/**
	 * Converts a Java timestamp to a Protobuf timestamp.
	 * @param javaTimestamp The Java timestamp to convert.
	 * @return The converted Protobuf timestamp.
	 */
	public static com.google.protobuf.Timestamp toProtobufTimestamp(Timestamp javaTimestamp) {
    if (javaTimestamp == null) {
      return null;
    }

    long millis = javaTimestamp.getTime();
    long seconds = millis / 1000;
    int nanos = (int) ((millis % 1000) * 1_000_000);
    return com.google.protobuf.Timestamp.newBuilder().setSeconds(seconds).setNanos(nanos).build();
  }

}
